src/rules/restrict-private-properties.ts

```diff
@@ -0,0 +1,111 @@
+import type { AST as JsonAST } from "jsonc-eslint-parser";
+
+import { fixRemoveObjectProperty } from "eslint-fix-utils";
+import * as ESTree from "estree";
+
+import { createRule } from "../createRule.ts";
+import { isJSONStringLiteral } from "../utils/predicates.ts";
+
+const defaultBlockedProperties = ["files", "publishConfig"];
+
+export const rule = createRule({
+	create(context) {
+		const blockedProperties =
+			context.options[0]?.blockedProperties ?? defaultBlockedProperties;
+
+		return {
+			"Program > JSONExpressionStatement > JSONObjectExpression"(
+				node: JsonAST.JSONObjectExpression,
+			) {
+				// Check if this is a private package
+				if (
+					!node.properties.some(
+						(property) =>
+							isJSONStringLiteral(property.key) &&
+							property.key.value === "private" &&
+							property.value.type === "JSONLiteral" &&
+							property.value.value === true,
+					)
+				) {
+					return;
+				}
+
+				// Check for blocked properties
+				for (const property of node.properties) {
+					if (
+						isJSONStringLiteral(property.key) &&
+						blockedProperties.includes(property.key.value)
+					) {
+						const isEmpty =
+							(property.value.type === "JSONArrayExpression" &&
+								property.value.elements.length === 0) ||
+							(property.value.type === "JSONObjectExpression" &&
+								property.value.properties.length === 0);
+
+						context.report({
+							data: {
+								property: property.key.value,
+							},
+							messageId: "unnecessaryProperty",
+							node: property,
+							...(
+								isEmpty
+									? {
+										fix: fixRemoveObjectProperty(
+											context,
+											property as unknown as ESTree.Property,
+										)
+									}
+									: {
+										suggest: [{
+											data: {
+												property: property.key.value,
+											},
+											fix: fixRemoveObjectProperty(
+												context,
+												property as unknown as ESTree.Property,
+											),
+											messageId: "unnecessaryProperty",
```

Comment:
The message for a suggestion should generally be the action that applying the suggestion will take.  For instance, this message should probably be something along the lines of `"Remove the {{name}} property"`.  That's what the IDE will show as the suggestion for this violation.